### DISCUSSION TITLE
Meldung "Warte auf Daten... (kann bis zu 10 Sekunden dauern)"

### DISCUSSION THREAD

**Discussion by nakott** [tbnobody/OpenDTU#2005] [General]
Hallo zusammen,
gestern bin ich von v24.1.21 auf v24.5.6 gegangen. Auch habe ich die Breaking Changes in 24.2.12 gelesen und bei mir bei der Benutzung der Web API angepasst. Soweit auch alles prima.
Wenn ich aber die OpenDTU Oberfläche per Web Browser zugreife, dann sind bei v24.1.21 alle Daten sofort angezeigt worden. Bei v24.5.6 bekomme ich jetzt erst diese Meldung, die unterschiedlich lange zwischen 1 und 5 Sekunden angezeigt wird.

![OpenDTUv24 5 6](https://github.com/tbnobody/OpenDTU/assets/22239367/30ad53e7-8f43-4637-9e36-02beaa4a19d9)

Was ist denn der technische Hintergrund dieser Wartezeit? Ich habe einen HM-1500. 
Vorab Danke und viele Grüße Dirk 

P.S: Ich denke, dass die Änderung hiermit kam. https://github.com/tbnobody/OpenDTU/commit/c27ecc3620f656fbdfaa7e9f0f3deca6a371a373
Kann es wirklich so viel Zeit extra benötigen, wenn je Inverter einen extra Aufruf benötigt? Wie gesagt, ich habe nur einen.

**Comment by Norbirad08** [comment]
Hallo,

bei mir geht diese Meldung (Warte auf Daten... (kann bis zu 10 Sekunden dauern))

nach dem Update überhaupt nicht mehr weg. Es wird zwar regelmäßig der Gesamtertrag insgesamt, heute und die Gesamtleistung aktualisiert aber keine Detaildaten der Phase 1 sowie der 4 Strings.

Was kann die Ursache dafür sein? 

In der xDTU App werden diese Daten weiterhin korrekt ausgegeben und aktualisiert.![image](https://github.com/tbnobody/OpenDTU/assets/95967491/533f41ce-93aa-494e-bba1-003c6cdcecf3)

**Reply by BeeJoo123** [reply]
Ich habe jetzt das selbe Problem, aber nur auf meinen Apple Geräten :-/
Unter Linux und Firefox läuft es mit allen Daten, echt seltsam. Woran kann das liegen ?

**Reply by BeeJoo123** [reply]
Ohje, es war nur der Werbe-Blocker /-)

**Reply by Marc9485** [reply]
@BeeJoo123 Liegt am ios update auf 26.1

**Reply by jblond4711** [reply]
Bei mir liegt es nicht am Werbeblocker sondern eindeutig an 26.1, nachdem ich wieder auf 26.01 gegangen bin geht es wieder.
Das Problem gab es schon einmal während der Beta Phase, weshalb Apple es jetzt wieder „kaputt“ gemacht hat bleibt deren Geheimnis.

**Reply by BeeJoo123** [reply]
Ohhje, stimmt! Auf MacOS 26.1 geht es, auf dem iPhone mit iOS 26.1 nicht. Herrjenochmal!

**Reply by Josef1966** [reply]
Gibt es dazu schon eine Lösung? Ich habe nämlich mit iPhone 14 und IOS 26.1 das gleiche Problem.
Auf dem Windows-PC funktioniert die Abfrage auf der Web-Oberfläche weiterhin sehr gut.

**Reply by TKlaube** [reply]
Es liegt nicht an IOS 26.1 , sondern eher an Safari unter IOS 26.1, denn mit Chrome unter IOS 26.1 funktioniert es richtig

**Reply by jblond4711** [reply]
Klappt es mit 26.2 jetzt wieder bei jemandem?

**Reply by TKlaube** [reply]
Nein, unter  26.2. ist es das gleiche Verhalten.

 Ich kann noch einen Beitrag leisten, ich habe zwei ESP32 die mit ESPUI ein WebUI erzeugen, auch das hängt unter Safari IOS 26.x

**Reply by Marc9485** [reply]
Mir ist auch noch aufgefallen, dass ich auf die shellys und tasmo nicht zugreifen kann über Safari

**Reply by TKlaube** [reply]
In diesen und ähnlichen Geräten stecken auch ESPxxx drinnen. vielleicht ist da irgendwo die Ursache?

**Comment by TKlaube** [comment]
<img width="414" height="896" alt="IMG_0292" src="https://github.com/user-attachments/assets/eaba3112-7155-4f15-a1ac-7710fe899236" />

**Comment by tbnobody** [comment]
Könnt ihr ggf. mal testen ob es mit Angabe des Ports funktioniert? also z.B. `http://<ip>:80`

Habe das gerade [hier](https://community.shelly.cloud/topic/10693-web-gui-not-working-in-safari-after-ios-26-public-beta-installation/) nachgelesen und bei mir klappt es damit.
Wieso, weshalb, warum? Keine Ahnung noch nicht.

**Reply by TKlaube** [reply]
Bei mir hilft Port 80 nicht!

**Reply by tbnobody** [reply]
Du hast recht. nachdem ich jetzt mehrfach die Seite neu geladen habe ist es jetzt wieder kaputt. (Wobei er mir in dem Fall jetzt den Port aus der Adressleiste löscht. Bin mir aber nicht sicher ob das die ersten male auch schon der Fall war). Habe in diversen Foren aber auch schon gelesen, dass es wohl ab und an funktioniert.

**Reply by tbnobody** [reply]
Ich schaue mir bald mal https://github.com/ESP32Async/ESPAsyncWebServer/pull/353 und https://github.com/tbnobody/OpenDTU/pull/2905 an um zu sehen ob das Problem dadurch behoben wird.

**Reply by TKlaube** [reply]
Bei mir ist die Portnummer auch gleich verschwunden, als die Ansicht erschien. Ja, ESP...WebServer ist vermutlich auch bei den Shelly's (basiert ja auch auf ESP) beteiligt und auch bei ESPUI. Trotzdem ist halt interessant, dass das Problem offensichtlich nur bei Safari ab 26.x auftritt.

**Reply by Marc9485** [reply]
@tbnobody Das ist ja lobenswert, dass du dir die Mühe machst aber eigentlich hat doch Apple es verkackt. Die Probleme haben nicht nur wir in DE wenn man mal so durch diverse Foren schaut. Und das so ein Konzern dass dann im 2. update nicht schafft zu ändern ist schon ein Witz

**Comment by tbnobody** [comment]
Some things I found out:
1. Its not working with the latest version of the webserver (3.9.3)
2. Safari does not even try to establish a connection (at least I don't see anything regarding "Websocket connected" on the serial console.
3. When using Vite as a Proxy (e.g. in the debug environment) it seems to work
4. ESPUI seems to use the same webserver as OpenDTU https://github.com/s00500/ESPUI/blob/ed9d7d58a45c2c3a2af4aa31aeb6078b7b228ec1/library.json#L18-L22
5. The issue also occours on Safari on Mac Books (so its not related to mobile devices) (Debug console shows `WebSocket connection to 'ws://192.168.30.79/livedata' failed: The network connection was lost.`
6. I was able to reproduce the issue with an old OpenDTU installation (24.2.18) (This version uses `mathieucarbou/ESP Async WebServer @ 2.7.0`)

Therefor I still think the issue is related to safari (also because it's working with all other browsers)

@mathieucarbou did you somehow hear similar issues elsewhere?
And to be honest, I am not quite sure whether it's related to the EspAsyncWebServer as the shellies seem to have the same problems. (Not sure what kind of webserver they are using)

**Reply by mathieucarbou** [reply]
Hi @tbnobody : please see here: https://github.com/hoylabs/OpenDTU-OnBattery/discussions/2331

This is reported since a while and seems to be an issue with the way Safari interacts with WebSocket under certain circumstances.
I have plenty of other apps using websocket working on Safari (like ESP-DASH), but OpenDTU, no. It is much more link I think to how the UI deals with the WS connection. If we activate some debug logs in the WebSocket layer, we see that Safari sends a close frame right after the open / pong.

Other people seem to have similar issue too on internet where the new ios broke some.

It does not seem related to EspAsyncWebServer... I wish it were in fact because we could have fixed it lol ;-) 

Although, there has been a fix in EspAsyncWebServer 3.9.3 regarding WS because in its new WS implementation, Safari can send pong replies fragmented (2 bytes in 1 frame, remaining in other frames), which is quite dumb and no other browser is doing that.

UI based on ESP-DASH (which I use a lot) are working fine so I guess this could be fixed on UI side by handling differently the WS connection maybe... But I have no other clue.

**Reply by mathieucarbou** [reply]
Note: to test you can uncomment some debug lines in AsyncWebSocket.cpp inside the _onData.

We also find a Safari bug once where Safari would send incomplete frames while closing...

I am pretty sure WS implementation in Safari is a mess and they are trying to refactor it, but considering the low investment Apple does in Safari, and also its inability in updating it faster than the OS like Chrome can, I think we all have to accept the fact that this browser is just too much trouble to support lol...

It really bothers me because I am only using Safari on mac, iphone, tablet to sync tabs and bookmarks and I then only use chrome for some specific dev tasks. Now I have to also use chrome for opendtu.

In any case, I also tested the main branch of EspAsyncWebServer in OpenDTU main branch because we did the release, in hope that it would fix the issue, but it did not 😞

**Reply by mathieucarbou** [reply]
Examples of closed connection while opening:

<img width="713" height="64" alt="image" src="https://github.com/user-attachments/assets/eba72079-7be2-408a-a7cd-462220cd2ba4" />
<img width="498" height="346" alt="image" src="https://github.com/user-attachments/assets/d7670206-3c1e-41c0-843c-6d452d450299" />
<img width="495" height="395" alt="image" src="https://github.com/user-attachments/assets/7c455784-717d-4916-b553-67c92c1e9c7c" />

When it works (for the console):

<img width="483" height="408" alt="image" src="https://github.com/user-attachments/assets/8dfc839f-3306-4fd9-9f40-fb285943a209" />

**Reply by tbnobody** [reply]
For me, the console doesn't work either. Or let me say, it works sometimes (not reproduceable)
On the other handy, the live view worked today at some time. (after closing and reopening safari, it was gone again)

I read a little bit in the webkit issue tracker today.... A lot of websocket related issues.

**Reply by mathieucarbou** [reply]
Yes... I don't think it is worth looking deeper in this issue until the Safari WebSocket implementation becomes stable again. At least not change any code that was working before.

**Reply by mathieucarbou** [reply]
And what I do not understand is why all apps I have based on ESP-DASH, which is 100% websocket-based, do work...

**Reply by okrause** [reply]
Hi, I have the same issue. Here two observations which might help you find the issue.
I am running MacOS 26.2 and latest OpenDTU. While OpenDTU shows data, inverter specific data is missing.
When I open Web Console in Safari, I see the following error message:
WebSocket connection to 'ws://admin:mypassword@opendtu.mydomain.internal/livedata' failed: The network connection was lost.

Interesting. I did not expect it to leak my login credentials ...

Anyway: When I logged out, it started to work and I was able to see my inverter data.

I also see this error, which looks unrelated:
"TypeError: window.crypto.randomUUID is not a function. (In 'window.crypto.randomUUID()', 'window.crypto.randomUUID' is undefined)"

**Comment by mathieucarbou** [comment]
There was a macOS and iOS update this morning with Safari 26.2 (`Version 26.2 (21623.1.14.11.9)`) .

The issue is still there though.

**Reply by TKlaube** [reply]
Apple is reportedly already working on version 26.3. Maybe this will fix the problem

**Reply by TKlaube** [reply]
Seems the problem is already otherwise known:

https://discussions.apple.com/thread/256142477?sortBy=rank

**Reply by mathieucarbou** [reply]
yes there are many reports.

But what is strange is that it works fine for example with ESP-DASH. I think it depends how the websocket connection is created.

**Example:**

page refresh with chrome: disco, then reco.

```
11:23:34.897 > D (144649) webapi: Websocket: [/livedata][3] disconnect
11:23:35.113 > D (144865) webapi: Websocket: [/livedata][4] connect
```

page refresh with safari: no logs. The websocket connection is closed before even reaching the server. So too me, this is an issue on client-side - the server is not even reached. And this is only happening with some UI frameworks.

So I suspect a timing issue.... Could be worth testing to setup the WS like this in OpenDTU, like in ESP-DASH:

```js
  function connect() {
    socket = new WebSocket(host);
    socket.onclose = onclose;
    socket.onopen = onopen;
    socket.onmessage = onmessage;
    socket.onerror = onerror;
  }
```

To fully initialize the object as fast as possible by using external functions, not anonymous ones that are created on the stack.

_edit_: just tried, no luck.

@tbnobody FYI I maybe found an issue in the heartCheck function... It can close the WS connection while in connecting or closing state:

```js
            this.heartInterval = setInterval(() => {
                if (this.socket.readyState === 1) {
                    // Connection status
                    this.socket.send('ping');
                } else {
                    this.initSocket(); // Breakpoint reconnection 5 Time
                }
            }, duration * 1000);
```

I suspect it should be something like:

```js
            this.heartInterval = setInterval(() => {
                if (this.socket.readyState === WebSocket.OPEN) {
                    // Connection status
                    this.socket.send('ping');
                } else if (this.socket.readyState !== WebSocket.CONNECTING) {
                    this.initSocket(); // Breakpoint reconnection 5 Time
                }
            }, duration * 1000);
```

Otherwise it only gives `duration` seconds to connect before forcing a close.

**Comment by FredHucht** [comment]
Same problem here with Safari under iOS 26.2. Firefox Klar works. 
However, dynamic updating works reproducibly after restarting Safari. Seems to be a problem with open websockets being closed on restart. Note that the power display in the Inverter tab is also affected.

**Comment by mathieucarbou** [comment]
FYI I also have some Shelly devices at home and Safari is also not working when opening them from their IP address.
Same issue: https://community.shelly.cloud/topic/10693-web-gui-not-working-in-safari-after-ios-26-public-beta-installation/

**Comment by jblond4711** [comment]
Hat Apple eigentlich noch vor das mal zu „reparieren“ ? Unfassbar wie ignorant mittlerweile der Support geworden ist. Es kommt nichtmal mehr eine Antwort von Apple auf mehrfaches Nachfragen.

**Comment by jblond4711** [comment]
Mit iPadOS 26.4 funktioniert es endlich wieder.

Edit: Das war wohl ein kurzer Glitsch in der Matrix denn jetzt geht es nicht mal mehr mit dem Firefox.

**Reply by mathieucarbou** [reply]
It works for me on Safari on MacOS with the new 26.4, but still not on my iPhone or iPad.
Tested Shelly + OpenDTU

**Reply by okrause** [reply]
Tested with iPadOS 26.4 and Safari on MacOS 26.4 to v25.9.11 and still no luck. It works occasionally, usually around the moments the inverter goes online/offline after it starts/stops when the panels start/stop delivering power.
But in normal operation, it still doesn't work for me.

**Comment by mathieucarbou** [comment]
This discussion relates to issue #2987 and I have proposed a fix in PR https://github.com/tbnobody/OpenDTU/pull/3004 that works fine.

**Comment by jblond4711** [comment]
Funktioniert bei jemandem das aktualisieren mit iPadOS 26.5 unter Safari? Mit iOS 26.5 funktioniert es komischerweise, zumindest auf dem getesteten iPhone 15pro.
